util/topsql/stmtstats/stmtstats.go

```diff
@@ -16,30 +16,75 @@ package stmtstats
 
 import (
 	"sync"
+	"time"
 
 	"go.uber.org/atomic"
 )
 
 var _ StatementObserver = &StatementStats{}
 
+// nowFunc returns the current time; it's overridden in tests.
+var nowFunc = time.Now
+
 // StatementObserver is an abstract interface as a callback to the corresponding
 // position of TiDB's SQL statement execution process. StatementStats implements
 // StatementObserver and performs counting such as SQLExecCount/SQLDuration internally.
 // The caller only needs to be responsible for calling different methods at the
 // corresponding locations, without paying attention to implementation details.
 type StatementObserver interface {
-	// OnExecutionBegin should be called before statement execution.
-	OnExecutionBegin(sqlDigest, planDigest []byte)
+	// OnDispatchBegin should be called on the begin of TiDB dispatch command.
+	OnDispatchBegin()
```

Comment:
I think you should explicitly point out protocol level things and other things. For protocol level things, I prefer to add a Cmd prefix, since Command is a unique concept that will not cause ambiguous:

- OnCmdDispatchBegin
- OnCmdDispatchFinish
- OnCmdQueryBegin
- OnCmdQueryFinish
- OnCmdQueryProcessStmtBegin
- OnCmdQueryProcessStmtFinish
- OnCmdStmtExecuteBegin
- OnCmdStmtExecuteFinish
- OnCmdStmtFetchBegin
- OnCmdStmtFetchFinish
- OnStmtReadyToExecute

This can be very clear that where these events should be called.